File: src-tauri/python/render_pdf.py
import os
import subprocess
from jinja2 import Environment, FileSystemLoader
import sys
import json

OUTPUT_DIR = os.path.join(os.path.dirname(os.getcwd()), "output")

def generate_pdf(data, template_file='template1.jinja', output_pdf_name='output.pdf', replace = False):
    """
    Renders a LaTeX Jinja template from the 'templates' folder and compiles it to PDF.
    """
    
    # Ensure the output directory exists
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)

    # 1. Create a Jinja2 environment, pointing to the 'templates' directory
    #    and set custom delimiters to avoid LaTeX curly-brace conflicts.
    env = Environment(
        loader=FileSystemLoader('python/templates'),
        block_start_string='((%',
        block_end_string='%))',
        variable_start_string='(((',
        variable_end_string=')))',
        comment_start_string='((#',
        comment_end_string='#))',
        trim_blocks=True,      # Automatically remove newline after a block
        lstrip_blocks=True 
    )

    # 2. Load the specified template
    template = env.get_template(template_file)

    # 3. Render the template with the provided data
    rendered_tex = template.render(**data)

    # 4. Write the rendered content to a temporary .tex file
    temp_tex = 'python/templates/temp_output.tex'
    with open(temp_tex, 'w', encoding='utf-8') as f:
        f.write(rendered_tex)

    # 5. Run pdflatex to compile .tex -> .pdf
    subprocess.run(['pdflatex', '-interaction=batchmode', 'temp_output.tex'], cwd="python/templates")

    # 6. Rename the generated PDF to your desired output name
    if os.path.exists('python/templates/temp_output.pdf'):
        destination_path = os.path.join(OUTPUT_DIR, output_pdf_name)
        if replace:
            os.replace('python/templates/temp_output.pdf', destination_path)
        else:
            new_file_name = create_new_file_name(output_pdf_name)
            destination_path = os.path.join(OUTPUT_DIR, new_file_name)
            os.rename('python/templates/temp_output.pdf', destination_path)

    # 7. Clean up auxiliary files if desired
    for ext in ['.aux', '.log', '.out', '.tex']:
        aux_file = 'python/templates/temp_output' + ext
        if os.path.exists(aux_file):
            os.remove(aux_file)

    return output_pdf_name


def create_new_file_name(new_pdf):
    """
    If `new_pdf` already exists, append (1), (2), etc. until we find a free filename.
    Then rename `original_pdf` to that new filename.
    """
    base, ext = os.path.splitext(new_pdf)
    candidate = new_pdf
    counter = 1

    while os.path.exists(candidate):
        candidate = f"{base}({counter}){ext}"
        counter += 1
    return candidate

if __name__ == '__main__':
    # Example data dictionary
    input_data = sys.stdin.read()
    print(OUTPUT_DIR)
    try:
        data_dict = json.loads(input_data)
        filename = f"{data_dict['name']}_cv.pdf" if data_dict['name'] else "new_cv.pdf"
        pdf_file = generate_pdf(data_dict, template_file='template1.jinja', output_pdf_name=filename, replace=True)
        response = {
            "status": "success",
            "data": f"Python received: {data_dict}, and generated PDF: {pdf_file}"
        }
        print(json.dumps(response))
        sys.stdout.flush()
    except Exception as e:
        print(json.dumps({"status": "error", "message": str(e)}))
        sys.stdout.flush()

File: src-tauri/src/render_pdf.rs
use serde_json::json;
use serde_json::Value;
use std::env;
use std::fs::{self, File};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::Command;
use tera::{Context, Tera};

/// Generate a PDF from a LaTeX template
pub fn generate_pdf(
    data: Value,
    template_file: &str,
    output_path: &str,
    replace: bool,
) -> String {

    let base_dir = env::current_dir().expect("Failed to get current directory");
    println!("📁 Current directory: {:?}", base_dir);
    let template_dir = base_dir.join("templates");
    match fs::read_dir(&template_dir) {
        Ok(entries) => {
            println!("📂 Folder contents:");
            for entry in entries {
                if let Ok(entry) = entry {
                    println!("  - {:?}", entry.file_name());
                }
            }
        }
        Err(_) => println!("❌ Could not read directory: {:?}", template_dir),
    }

    // Load and render the LaTeX template using Tera (Jinja2-like)
    let template_path = format!("{}/{}", template_dir.display(), "*.jinja");
    let tera = match Tera::new(&template_path) {
        Ok(t) => t,
        Err(e) => panic!("Template parsing error: {}", e),
    };

    println!(
        "📜 Loaded templates: {:?}",
        tera.get_template_names().collect::<Vec<_>>()
    );

    let mut context = Context::new();
    for (key, value) in data.as_object().unwrap() {
        context.insert(key, value);
    }

    let rendered_tex = tera
        .render(template_file, &context)
        .expect("Failed to render template");

    // Save the rendered LaTeX content to a .tex file

    let temp_tex_path = template_dir.join("temp_output.tex");

    let mut tex_file = File::create(&temp_tex_path).expect("Failed to create .tex file");
    tex_file
        .write_all(rendered_tex.as_bytes())
        .expect("Failed to write .tex file");

    if !temp_tex_path.exists() {
        panic!(
            "❌ Error: LaTeX source file {:?} does not exist!",
            temp_tex_path
        );
    }

    // Run pdflatex to generate the PDF
    let pdflatex_status = Command::new("pdflatex")
        .arg("-interaction=batchmode")
        .arg("temp_output.tex")
        .current_dir(&template_dir)
        .status()
        .expect("Failed to execute pdflatex");

    if !pdflatex_status.success() {
        panic!("pdflatex failed to compile the document.");
    }

    // Rename the generated PDF
    let temp_pdf_path = template_dir.join("temp_output.pdf");
    let destination_pdf_path = output_path;
    if temp_pdf_path.exists() {
        fs::rename(&temp_pdf_path, &destination_pdf_path).expect("Failed to rename PDF file");
    }

    for ext in [".aux", ".log", ".out", ".tex"] {
        let aux_file = template_dir
            .join("temp_output")
            .with_extension(ext.trim_start_matches('.'));

        if aux_file.exists() {
            match fs::remove_file(&aux_file) {
                Ok(_) => println!("🗑️ Deleted {:?}", aux_file),
                Err(e) => println!("⚠️ Failed to delete {:?}: {}", aux_file, e),
            }
        }
    }

    return destination_pdf_path.to_string();
}

// /// Generate a new filename if a file already exists
// fn create_new_file_name(original_pdf: &str) -> String {
//     let output_dir = std::env::current_dir()
//         .expect("Failed to get current directory")
//         .parent()
//         .expect("Failed to get parent directory")
//         .join("output");
//     let base_name = original_pdf.trim_end_matches(".pdf");

//     let mut file_name = original_pdf.to_string();
//     let mut candidate = output_dir.join(&file_name);
//     let mut counter = 1;

//     while candidate.exists() {
//         file_name = format!("{}({}).pdf", base_name, counter);
//         candidate = output_dir.join(&file_name);
//         counter += 1;
//     }
//     file_name
// }

/// Entry point to process JSON input
pub fn process_pdf_request(json_data: &str, output_path: &str) -> String {
    let data: Value = serde_json::from_str(json_data).expect("Failed to parse JSON");
    println!("📥 Received Path: {}", output_path);
    let template = data["template"].as_str().unwrap_or("template1");
    let template_name = format!("{}.jinja", template);
    println!("{}", data);
    let pdf_file = generate_pdf(data, &template_name, &output_path, false);

    let response = json!({
        "status": "success",
        "data": format!("Generated PDF: {}", pdf_file)
    });

    serde_json::to_string(&response).expect("Failed to serialize response")
}
